apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -309,3 +321,106 @@ class FrameworkError {
   message: string;
   name: string;
 }
+
+function sanitizeControlValuesByOutputSchema(
+  controlValues: Record<string, unknown>,
+  type: StepTypeEnum
+): Record<string, unknown> {
+  const outputSchema = channelStepSchemas[type].output || actionStepSchemas[type].output;
+
+  if (!outputSchema || !controlValues) {
+    return controlValues;
+  }
+
+  const ajv = new Ajv({ allErrors: true });
+  addFormats(ajv);
+  const validate = ajv.compile(outputSchema);
+  const isValid = validate(controlValues);
+  const errors = validate.errors as null | ErrorObject[];
+
+  if (isValid || !errors || errors?.length === 0) {
+    return controlValues;
+  }
+
+  return replaceInvalidControlValues(controlValues, errors);
+}
+
+/**
+ * Fixes invalid control values by applying default values from the schema
+ *
+ * @example
+ * // Input:
+ * const values = { foo: 'invalid' };
+ * const errors = [{ instancePath: '/foo' }];
+ * const defaults = { foo: 'default' };
+ *
+ * // Output:
+ * const fixed = { foo: 'default' };
+ */
+function replaceInvalidControlValues(
+  normalizedControlValues: Record<string, unknown>,
+  errors: ErrorObject[]
+): Record<string, unknown> {
+  const fixedValues = _.cloneDeep(normalizedControlValues);
+
+  errors.forEach((error) => {
+    const path = getErrorPath(error);
+    const defaultValue = _.get(previewControlValueDefault, path);
+    _.set(fixedValues, path, defaultValue);
+  });
+
+  return fixedValues;
+}
+
+/*
+ * Extracts the path from the error object:
+ * 1. If instancePath exists, removes leading slash and converts remaining slashes to dots
+ * 2. If no instancePath, uses missingProperty from error params
+ * Example: "/foo/bar" becomes "foo.bar"
+ */
+function getErrorPath(error: ErrorObject): string {
+  return (error.instancePath.substring(1) || error.params.missingProperty).replace(/\//g, '.');
+}
+
+const EMPTY_STRING = '';
+const WHITESPACE = ' ';
+const DEFAULT_URL_TARGET = '_blank';
+const DEFAULT_URL_PATH = 'https://www.redirect-example.com';
+const DEFAULT_TIP_TAP_EMPTY_PREVIEW: TipTapNode = {
+  type: 'doc',
+  content: [
+    {
+      type: 'paragraph',
+      attrs: {
+        textAlign: 'left',
+      },
+      content: [
+        {
+          type: 'text',
+          text: EMPTY_STRING,
+        },
+      ],
+    },
+  ],
+};
+
+/**
+ * Default control values used specifically for preview purposes.
+ * These values are designed to be parsable by Liquid.js and provide
+ * safe fallback values when generating preview.
+ */
+export const previewControlValueDefault = {
```

Comment:
Following on the previous comment, these defaults should probably belong to the framework so that preview is more graceful in code-first workflows.

Reply:
I'd argue that having defaults on the schema makes the flow more unpredictable. let's discuss about the best solution.